Implement Sweep V0 for testing GC Sweep with shorter timers

markfields opened the pull request:
## Description

The next step in implementing/rolling out GC is to add a way for the team to Dogfood full GC Sweep, with much shorter timeouts than the typical 30 days.  This is called "Sweep V0".  When Sweep V0 is enabled, there will be ways to wrongly get content swept (e.g delete something then wait longer than the timeout), so beware of this.

But it lets us dogfood Sweep, both flushing out bugs in Container/Component code with how they reference objects, and also to see what the user/telemetry experience is when swept content is still referenced.

## Reviewer Guidance

Still need to add tests, and think about if there are any other changes required (e.g. in Summarizer)

markfields committed:
Implement Sweep V0

vladsud commented on packages/runtime/container-runtime/src/garbageCollection.ts:
```diff
@@ -365,6 +365,8 @@ export class GarbageCollector implements IGarbageCollector {
     public readonly trackGCState: boolean;
 
     private readonly testMode: boolean;
+    private readonly sweepV0TimeoutMs: number | undefined;
```
Not sure I see a value in this member, given that value from feature gate overwrites this.sweepTimeoutMs and that's how it gets used.

vladsud replied:
Ah, did not notice sweepV0. Maybe break it into more targeted properties, like deleteUnusedRoutesImmidiatly() (to be used in sweepIfEnabled()

vladsud commented on packages/runtime/container-runtime/src/garbageCollection.ts:
```diff
@@ -514,6 +521,11 @@ export class GarbageCollector implements IGarbageCollector {
 
+        // Sweep V0 overrides Sweep Timeout, even setting it if SessionExpiry is disabled
+        if (this.sweepV0TimeoutMs !== undefined) {
+            this.sweepTimeoutMs = this.sweepV0TimeoutMs;
```
Should this be a Math.min() ? Just in case one value is much larger than another one.

vladsud commented on packages/runtime/container-runtime/src/garbageCollection.ts:
```diff
@@ -490,7 +492,12 @@ export class GarbageCollector implements IGarbageCollector {
 
+        // Sweep V0 is a test mode for using a very short timeout in order to enable testing of real sweep behavior.
+        // It's similar to testMode but rather than deleting immediately, it exercises the Sweep timers and codepaths
+        this.sweepV0TimeoutMs = this.mc.config.getNumber("Fluid.GarbageCollection.TestOverride.SweepV0TimeoutMs");
```
I think the usage of this.sweepTimeoutMs across the code is wrong.
Sweep duration should be the same for all collaborators in the same document for system to work correctly. Even hard coding the value in the code does not provide that guarantee (it may change in next version), using feature gates make it even worse. I think the best course of action for this value (from code or feature gate) to be used to be written into file metadata, and every client using same value from metadata for lifetime of the file. Allowing to increase this value (in next summaries), but never decrease.

It's not ideal (test files are "contaminated" with small timeout forever), but I think we can work with it (whoever creates such files should put a big warning at the top of the file that it's for test purposes - it required anyway, as uneducated user opening this file can run into troubles).

tyler-cai-microsoft replied:
This reminds me, we should have some way of being able to ensure this value is consistent across clients. It feels like this needs to be an op of some sorts. A half-baked thought would be to create a GC datastore. We would be using our own datastore logic, and we could definitely take advantage of a significant portion of the logic used to sync ops/datastores. It would help us solve the problem of setting sessionExpiry and modifying it. It would help us solve the problem of having inconsistent sweep timings. It feels like the GC datastore would be a special datastore. Granted I do not know what the cost of doing something like this is.

Pros:
- resuses a lot of logic to ensure consistency across clients without having to rewrite everything
- allows us to take advantage of any sort of performance improvements to datastores and the DDS we use
- potentially more recovery abilities, might be able to leverage rollback capabilities from certain DDSs, so we have the potential to rollback without having to design it ourselves as a GC team.
- provides an example to others of how to use Datastores (may not be a con if we don't want others to do this)
- may allow us to remove a bunch of GC specific summary things.

Cons:
- One issue is that garbage collection would be creating its own ops. 
- The cost of doing this vs. slowly solving those problems over time.
- We may be exposing functionality that others may use to write their own GC algorithms
- All clients would have to process the GC ops.

markfields replied:
Thanks for pointing this out, both of you. My next step in this PR is to address this.  See also the doc I shared about adjusting "Sweep V0" plan to "Hyper Sweep Mode" where all relevant timers are shortened together to keep the overall flow consistent.  Planning to implement that here, but gotta figure out about keep these timers locked per file first.

markfields replied:
Ok in latest iteration if we use a TestMode property in the metadata to stamp the file.

markfields commented on packages/runtime/container-runtime/src/garbageCollection.ts:
```diff
@@ -544,8 +556,9 @@ export class GarbageCollector implements IGarbageCollector {
-        // Override inactive timeout if test config or gc options to override it is set.
-        this.inactiveTimeoutMs = this.mc.config.getNumber("Fluid.GarbageCollection.TestOverride.InactiveTimeoutMs") ??
+        this.inactiveTimeoutMs =
+            this.sweepV0 ? 0 : // For SweepV0 consider unreferenced objects immediately inactive
```
0 ?? 1 === 0 I thought

tyler-cai-microsoft replied:
Might be better to make this a separate line to show that you're overriding the inactiveTimeoutMs

vladsud commented on packages/runtime/container-runtime/src/garbageCollection.ts:
```diff
@@ -555,7 +568,9 @@ export class GarbageCollector implements IGarbageCollector {
-        this.testMode = this.mc.config.getBoolean(gcTestModeKey) ?? this.gcOptions.runGCInTestMode === true;
+        this.testMode =
+            !this.sweepV0 && // Disable testMode if we're doing Sweep V0
+            (this.mc.config.getBoolean(gcTestModeKey) ?? this.gcOptions.runGCInTestMode === true);
```
It feels like you want to imply Fluid.GarbageCollection.TestOverride.SweepV0TimeoutMs => gcTestModeKey
And other places where you test for presence of this.sweepV0TimeoutMs should be actually testing for presence of test mode.
If so, maybe create a tiny wrapper around this.mc.config.getBoolean(gcTestModeKey) that would return true if the other feature gate is set, but otherwise the rest of the code should use SweepV0TimeoutMs only to overwrite this.sweepTimeoutMs?

markfields replied:
Yeah, this kind of stuff needs to be better encapsulated.  I opened https://dev.azure.com/fluidframework/internal/_workitems/edit/1361 about this and I think it covers your feedback.

vladsud commented on packages/runtime/container-runtime/src/garbageCollection.ts:
```diff
@@ -787,9 +802,9 @@ export class GarbageCollector implements IGarbageCollector {
-        this.logSweepEvents(logger, currentReferenceTimestampMs);
+        // If enabled, run Sweep to delete objects that have been unreferenced sufficiently long
+        // This also logs once per node when it is (or  would be) swept
+        this.sweepIfEnabled(logger, currentReferenceTimestampMs);
```
Shouldn't this code run only when summary is generated? I understand the desire to go faster, but will it cause trouble? Maybe not, but worth considering all the interesting cases.
It sounds like it's created in every container, including non-summarizing container. This user may not have write permissions to a file. Will it result in some errors? Or all the modifications we do here results in no ops and only reflect back to file through summarizer?

markfields replied:
It does only run when the summary is generated, if you look at its call site and keep tracing back.

markfields replied:
But yes, this class is instantiated for all clients. Maybe we should think about if this is necessary and/or expensive.

markfields replied:
Actually it's kind of nice that it does, because then I can do things like throw an error to close the main container if we try to open a test mode document we don't understand (see top of configForExistingContainer function)

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/garbageCollection.ts:
```diff
@@ -1260,6 +1277,11 @@ export class GarbageCollector implements IGarbageCollector {
             }
 
+            const age = currentReferenceTimestampMs - nodeStateTracker.unreferencedTimestampMs;
+            idsToSweep.add({ nodeId, nodeType, age });
+        });
+
+        idsToSweep.forEach(({ nodeId, nodeType, age }) => {
```
Do we need this code here?

markfields replied:
I think continuing to log when we delete stuff will be useful.  I'll probably refactor this back into the upper forEach loop to make it simpler.

tyler-cai-microsoft replied:
I specifically meant those lines, of code 1281 - 1284, if you removed those lines, wouldn't the code still work the same?

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/garbageCollection.ts:
```diff
@@ -1270,12 +1292,16 @@ export class GarbageCollector implements IGarbageCollector {
         });
+
+        if (this.sweepV0) {
+            this.runtime.deleteUnusedRoutes(Array.from(idsToSweep).map(({ nodeId }) => nodeId));
```
We will need an end-to-end test to verify that we can actually reach here

markfields commented on packages/runtime/container-runtime/src/garbageCollection.ts:
```diff
@@ -1240,16 +1255,18 @@ export class GarbageCollector implements IGarbageCollector {
             return;
         }
 
+        const idsToSweep = new Set<{ nodeId: string; nodeType: string; age: number; }>();
```
Note: This doesn't work! Set does not do deep equality.

markfields committed:
simplify sweepIfEnabled

markfields committed:
giut

markfields committed:
Move storage policy definition/calculation up to Document Service Factory

This lets us pass it via constructor into both EpochTracker and OdspDocumentStorageService

markfields committed:
Merge branch 'main' into pr/markfields/11084-2

markfields committed:
Revert changes covered by PR #11168

markfields committed:
Override epochTracker snapshot expiry via config setting

markfields committed:
Revert attempt at disabling snapshot expiry in driver

markfields committed:
Add initial tests for GC Configuration

markfields committed:
Initial gcConfig refactor

markfields committed:
Implement SweepV0 test config

And general GC Test Config infra

markfields committed:
Properly use SweepV0 status in gc.ts

Move all session-scoped config back to garbageCollection.ts

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/gcConfig.ts:
```diff
@@ -0,0 +1,152 @@
+    // Crash if we see a test mode we don't recognize, since we don't know how it's implemented or should be used
+    const testMode = metadata?.gcTestMode;
+    if (testMode !== undefined && testMode !== "SweepV0") {
+        unreachableCase(testMode, `Cannot open container created under unknown GC Test Mode [${testMode}]`);
```
do we have a way to get out of this situation?

markfields replied:
This is by design. The point is to only allow editing a test document by code that knows about it. In other words, any test mode document is basically considered "corrupted" to clients not aware of how to use it. It's a fail-fast mechanism; a test mode could result in a legit corruption anyway.

With 20-20 hindsight I would make this same change a few versions ago to prevent SweepV0 docs from being opened by older clients.

markfields replied:
See #11225

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/gcConfig.ts:
```diff
@@ -0,0 +1,152 @@
+            sweepAllowed: false,
+            gcAllowed,
+            prevSummaryGCVersion,
+        };
```
We want to return SessionExpiryTimeoutMs if it is at some value.

markfields replied:
I think not, since it's irrelevant if sweep is not allowed for the file.  But do you think we should enforce sessionExpiry even if sweep is not allowed?

markfields replied:
Ok I agree, I switched it to be independent of sweep.  Bunch of tests broke too which is good :)

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/gcConfig.ts:
```diff
@@ -0,0 +1,152 @@
+
+    const sweepV0: boolean = testMode === "SweepV0";
+    const snapshotCacheExpiryMs = sweepV0 ? 0 : defaultCacheExpiryTimeoutMs; // Ignore snapshot expiry for SweepV0
+    const sweepTimeoutBufferMs = metadata.sweepTimeoutBufferMs ?? defaultBufferMs; // For SweepV0 we expect it to be set
```
if this is not set, we should not run sweepV0

markfields replied:
Yeah, I'm on the fence between overly-asserting the relationship between different settings v. just write the code to do the right thing and log enough to diagnose if something weird happens.  Code elsewhere (I think not written yet!) will in fact always write sweepTimeoutBufferMs if we're in SweepV0 mode.

I can add an assert if you think that's a better approach.  I don't think it will work well to disable SweepV0 if this is missing though - Likely other timeouts will be small, and we should keep the file marked as test-mode to keep it from leaking out to production environment.

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/test/runningSummarizer.spec.ts:
```diff
@@ -32,7 +32,8 @@ class MockRuntime {
 
-describe("Runtime", () => {
+//* SKIP
+describe.skip("Runtime", () => {
```
Why are we skipping this?

markfields replied:
Kian accidentally checked in a `.only` in another test file.  When I removed it, a bunch of these tests fail.  I told him about it and he's going to fix it, but this unblocks me.

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:
```diff
@@ -114,6 +114,113 @@ describe("Garbage Collection Tests", () => {
     });
 
+    //* ONLY
+    describe.only("Configuration", () => {
```
only?

markfields commented:
Note to self -- Here are some todo's as I'm reviewing this again myself:
- [ ] Make sure we write all persisted props to metadata (e.g. buffer)
- [ ] Think about what older client will do if it opens a SweepV0 doc (can we put something in that will crash it...?)
- [ ] e2e test(s)
- [ ] Double-check unit test coverage of diff
- [ ] Move testMode up to all props, not just when sweepAllowed
- [ ] //* comments
- [ ] Maybe: replace prevSummaryGCVersion with (non-readonly) latestSummaryGCVersion in GcContainerConfig
- [ ] Add proper options key for session expiry
- [ ] Update GarbageCollectorLoaded log

markfields committed:
PR Feedback

markfields committed:
Write to metadata new persisted values

markfields committed:
Set sessionExpiryTimeoutMs regardless of sweep status

markfields committed:
Decoupline sweepAllowed from timeouts

And get rid of default snapshot expiry,
instead comparing metadata value with provided value

markfields committed:
Fix tests

markfields committed:
effectiveSessionExpiryTimeoutMs for this session's expiry

markfields committed:
Move snapshotCacheExpiryMs handling out of gcConfig.ts

Simpler now!

markfields committed:
Remove final //* todo comments

markfields commented:
I have to close this, it's not converging and I'll be out of the office the next 2 weeks. I'll pick it up when I return.

I am going to open at least one smaller PR from some of the progress I made here, e.g. for adding UTs, maybe some of the config stuff.
